[AMDGPU] Update PromoteAlloca to handle GEPs with variable offset.

sgundapa opened the pull request:
In case of variable offset of a GEP that can be optimized out, promote alloca is updated to use the refereshed index to avoid an assertion.

Issue found by fuzzer.

sgundapa committed:
[AMDGPU] Update PromoteAlloca to handle GEPs with variable offset.

In case of variable offset of a GEP that can be optimized out, promote
alloca is updated to use the refereshed index to avoid an assertion.

Issue found by fuzzer.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -385,13 +385,42 @@ static bool isSupportedMemset(MemSetInst *I, AllocaInst *AI,
          match(I->getOperand(2), m_SpecificInt(Size)) && !I->isVolatile();
 }
 
+static bool hasVariableOffset(GetElementPtrInst *GEP) {
```
This looks like you're pre-filtering for a scenario that could happen in the map lookup. You should directly detect when this delete happens rather than assuming the set of cases it could

sgundapa replied:
It will make the code future proof. Will make the change

arsenm replied:
One possible solution is to track these values with WeakVH. Then you'll just have to check if the value was already deleted before deleting it. This appears to be how SROA handles it, (see how DeadInsts is a vector of WeakVH)

sgundapa replied:
WeakVH might not be useful here as the Value still exists in the code. We can do a DCE but I am not sure if it guaranteed if the Value has no uses in the code.

arsenm replied:
You can easily check if the operation is dead. This pass is essentially doing the same thing as SROA, which checks isInstructionTriviallyDead

arsenm commented on llvm/test/CodeGen/AMDGPU/promote-alloca-array-aggregate.ll:
```diff
@@ -122,6 +122,34 @@ define amdgpu_vs void @promote_load_from_store_aggr() #0 {
+  %G1 = getelementptr inbounds i8, ptr addrspace(5) %f1, i32 8
+  %foo3.unpack2 = load i32, ptr addrspace(1) getelementptr inbounds (i8, ptr addrspace(1) @block4, i64 8), align 4
+  store i32 %foo3.unpack2, ptr addrspace(5) %G1, align 4
+  %foo4 = load i32, ptr addrspace(5) %G1, align 4
```
give a better name to indicate this is the interesting value

arsenm commented on llvm/test/CodeGen/AMDGPU/promote-alloca-array-aggregate.ll:
```diff
@@ -122,6 +122,34 @@ define amdgpu_vs void @promote_load_from_store_aggr() #0 {
+
+; This should should not crash on variable offset that can be
+; optimized out (variable foo4 in the test)
+define amdgpu_vs void @promote_load_from_store_aggr_varoff() local_unnamed_addr {
```
```suggestion
define amdgpu_vs void @promote_load_from_store_aggr_varoff() {
```

sgundapa replied:
Sure

sgundapa committed:
[AMDGPU] Update PromoteAlloca to handle updated GEP Indexes.

The PromoteAlloca pass was using outdated cached GEP indices in some cases,
leading to an assertion failure when encountering aliased GEP indices that could
be optimized out. This commit fixes the issue by refreshing the cached index
before use.

Issue found by fuzzer.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -459,6 +422,36 @@ static Value *GEPToVectorIndex(GetElementPtrInst *GEP, AllocaInst *Alloca,
 }
 
+// Function to check if a Value is an operand of a GetElementPtrInst.
+static bool isValueInGEP(GetElementPtrInst *GEP, Value *ValueToCheck) {
```
This still feels like too specific of a check, and the same form of pattern could arise from other instructions. It would be better to detect it as part of the transformation

sgundapa committed:
Merge branch 'llvm:main' into pro_alloc_crash

sgundapa committed:
[AMDGPU] Update PromoteAlloca to handle changed GEP Indexes.

The PromoteAlloca pass was using outdated cached GEP indices in some cases,
leading to an assertion failure. This commit fixes the issue by using ValueHandle
to track the change in values.

Issue found by fuzzer.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp:
```diff
@@ -385,16 +385,21 @@ static bool isSupportedMemset(MemSetInst *I, AllocaInst *AI,
   assert(I != GEPIdx.end() && "Must have entry for GEP!");
-  return I->second;
+
+  if (Value *IndexValue = I->second)
+    return IndexValue;
+
+  llvm_unreachable(
+      "Index value is not present in Cached GEPs! This indicates a "
+      "bug in the pass that populates GEPIdx.");
```
```suggestion
    Value *IndexValue = I->second;
    assert(IndexValue && "index value missing from GEP index map");
    return IndexValue;
```

sgundapa committed:
Update llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

arsenm committed:
clang-format fix
